tensorflow_cloud/containerize.py

```diff
@@ -39,209 +43,334 @@
+
+        Raises:
+            RuntimeError: if there are any errors when building or publishing a
+            docker image.
+        """
+        for line in logs_generator:
+            try:
+                unicode_line = line.decode('utf-8').strip()
+                logger.info(unicode_line)
+            except UnicodeError:
+                logger.warning('Unable to decode logs.')
+            try:
+                line = json.loads(unicode_line)
+                if line.get('error'):
+                    raise RuntimeError(
+                        'Docker image {} failed: {}'.format(name, str(
+                            line.get('error'))))
+            except json.decoder.JSONDecodeError:
+                raise RuntimeError(
+                    'There was an error decoding the Docker logs')
 
-def _generate_name(docker_registry):
-    """Returns unique name+tag for the docker image."""
-    unique_tag = uuid.uuid4()
-    return '{}/{}:{}'.format(
-        docker_registry, 'tensorflow-train', unique_tag)
 
+class GcloudContainerBuilder(ContainerBuilder):
+    """Container builder that uses Google cloud build."""
 
-def _get_logs(logs_generator, name):
-    """Decodes logs from docker and generates user friendly logs.
+    def get_docker_image(self):
+        """Builds, publishes and returns a docker image."""
+        self._get_tar_file_path()
+        storage_object_name = self._upload_tar_to_gcs()
+        image_uri = self._generate_name()
 
-    Args:
-        logs_generator: Generator returned from docker build/push APIs.
-        name: String, 'build' or 'publish' used to identify where the generator
-            came from.
+        logger.info('Building and publishing docker image using Google '
+                    'Cloud Build: {}'.format(image_uri))
+        build_service = discovery.build(
+            'cloudbuild', 'v1', cache_discovery=False)
+        request_dict = self._create_cloud_build_request_dict(
+            image_uri, storage_object_name)
 
-    Raises:
-        RuntimeError: if there are any errors when building or publishing a
-        docker image.
-    """
-    for line in logs_generator:
-        try:
-            unicode_line = line.decode('utf-8').strip()
-            logger.info(unicode_line)
-        except UnicodeError:
-            logger.warning('Unable to decode logs.')
         try:
-            line = json.loads(unicode_line)
-            if line.get('error'):
+            # Call to queue request to build and push docker image.
+            create_response = build_service.projects().builds().create(
+                projectId=self.project_id,
+                body=request_dict
+            ).execute()
+
+            # Create returns a long-running Operation
+            # https://cloud.google.com/cloud-build/docs/api/reference/rest/v1/operations#Operation
+            # This contains the build id, which we can use to get the status.
+
+            timeout = time.time() + 60*5   # 5 minutes from now
```

Comment:
The timeout value should probably be configurable

Reply:
Removed timeout and added max attempts flag.